prover/src/constraints/evaluator/logup_gkr.rs

```diff
@@ -68,90 +70,184 @@ where
-                        );
-                    let inv_divisor = trans_constraints_divisors
-                        .get_inverse_divisor_eval(trans_constraint_idx, step);
-
-                    combined_evaluations += numerator * inv_divisor;
-                }
-
-                // combine boundary constraints
-                {
-                    let boundary_numerator = self
-                        .lagrange_kernel_constraints
-                        .boundary
-                        .evaluate_numerator_at(&lagrange_frame);
-
-                    combined_evaluations += boundary_numerator * boundary_divisors_inv[step];
-                }
 
-                combined_evaluations
-            };
-
-            // compute and combine the transition constraints for the s-column.
-            // The s-column implements the cohomological sum-check argument of [1] and
-            // the constraint we enfore is exactly Eq (4) in Lemma 1 in [1].
-            //
-            // [1]: https://eprint.iacr.org/2021/930
-            let s_col_combined_evaluation = {
-                trace.read_main_trace_frame_into(step << lde_shift, &mut main_frame);
-                trace.read_aux_trace_frame_into(step << lde_shift, &mut aux_frame);
-
-                let l_cur = aux_frame.current()[l_col_idx];
-                let s_cur = aux_frame.current()[s_col_idx];
-                let s_nxt = aux_frame.next()[s_col_idx];
-
-                evaluator.build_query(&main_frame, &mut query);
-                let batched_query = self.gkr_data.compute_batched_query(&query);
-
-                let rhs = s_cur - mean + batched_query * l_cur;
-                let lhs = s_nxt;
-
-                let divisor_at_step =
-                    s_col_constraint_divisor[step % (domain.trace_to_ce_blowup())];
-
-                (rhs - lhs) * self.s_col_composition_coefficient.mul_base(divisor_at_step)
-            };
-
-            combined_evaluations_acc[step] +=
-                lagrange_combined_evaluations + s_col_combined_evaluation;
+        #[cfg(feature = "concurrent")]
+        combined_evaluations_acc
+            .par_iter_mut()
+            .enumerate()
+            .fold(
+                || {
+                    (
+                        LagrangeKernelEvaluationFrame::new_empty(),
+                        EvaluationFrame::new(trace.trace_info().main_segment_width()),
+                        EvaluationFrame::new(trace.trace_info().aux_segment_width()),
+                        vec![E::BaseField::ZERO; evaluator.get_oracles().len()],
+                    )
+                },
+                |(mut lagrange_frame, mut main_frame, mut aux_frame, mut query), (step, acc)| {
+                    // compute Lagrange kernel frame
+                    trace.read_lagrange_kernel_frame_into(
+                        step << lde_shift,
+                        l_col_idx,
+                        &mut lagrange_frame,
+                    );
```

Comment:
One potential thing to check is whether using `Vec::push()` inside this method is slowing things down for some reason. It shouldn't but also we usually use direct element assignment for things like this.